Assets/Scripts/CombatBehaviour.cs

```diff
@@ -0,0 +1,21 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class CombatBehaviour
+{
+    public MoveBehaviour moveBehaviour;
+    
+    public abstract void Act();
+}
```

Comment:
For reference I think the two classes `CombatBehaviour` and `MoveBehaviour` should be disentagled fully. I don't `CombatBehaviour` should _rely_ on `MoveBehaviour`.

Reply:
lol I'm gonna need an enemy more complex than this Bomb before we can even test an implementation then. It literally doesn't have a CombatBehaviour distinct from its MoveBehaviour under your suggested framework, it just closes distance.

Reply:
> lol I'm gonna need an enemy more complex than this Bomb before we can even test an implementation then. It literally doesn't have a CombatBehaviour distinct from its MoveBehaviour under your suggested framework, it just closes distance.

I know, I expect that they'll both have the same functionality right now. It's mainly a matter of API design.

Reply:
I'm going to disentangle them by renaming CombatBehaviour to AttackBehaviour, because at a guess, that's more what you were thinking. That being said, I'm also going to remove "JustMove" as an AttackBehaviour because it doesn't really make sense, that's not an attack.

Reply:
So here's my thinking. We can essentially factor generic AI into three phases:

1. Idle: no information on active threats, be it player or mob
2. Aware: is aware of a threat, and is either avoiding or seeking
3. Engaged: is directly engaged with threat, and is either acting offensively/defensively

The most trivial rule is just "do nothing" for all three. A _slightly_ less trivial rule is to do nothing for `idle`, move directly towards enemy in `aware` and move directly towards enemy in `engage`. That's the behavior our sample enemies -- the bombs -- display right now. For debugging purposes, it's sufficient to use `Debug.Log` to confirm that we are in the correct phases during a gameplay test. I think we need to get comfortable using gameplay tests and debug logs to confirm our implementations.

Each of these, in my opinion, should be as independent as possible. There may be some which have very limited interplay (e.g. idle behavior A is _only_ used with aware behaviors B or C and engage behavior D...) but from a code-organization perspective it should still be factored separately because in the general case I think we expect reusability of these behaviors.

Alternatively, if we really do expect that each monster is going to be unique, we can simply have methods `Combat{Idle, Aware, Engage}' and reimplement them in each `Enemy` subclass, factoring out reusable code (e.g. direct movement as we currently have) into a combat-utility namespace (similarly to how we handle angle computation).

What do you think?

Reply:
Okay, the `idle`, `aware`, `engaged` trichotomy makes a lot more sense than just `MoveBehaviour` and `AttackBehaviour` dichotomy to me, especially since movement is frequently part of attacking, and `engaged` is going to look like fleeing for some enemies. I also don't think we should expect that each monster is going to be wholly unique, I'm very sure we're going to be able to make `idle` behaviours that can work for different enemies just by tweaking a couple of instance variables.

I do wonder though if it makes sense to have three different abstract classes for these kinds of behaviours though. Like we clearly have a case with the `Bomb` where an enemy has the same `aware` behaviour as their `engaged` behaviour, and it seems like it might just make more sense to make an abstract `EnemyBehaviour` class that all these behaviours extend and then give the abstract `Enemy` class 3 `EnemyBehaviour` instance variables, say `idleAction`, `awareAction`, and `engagedAction`. We could make an `enum ENEMY_STATE` that takes on values `idle`, `aware`, and `engaged`, then give the `Enemy` abstract class an instance variable storing its current state from the 3 `ENEMY_STATE` options and a basic `Update` implementation that updates its state variable (maybe using an abstract method `DetermineState` whose implementation varies by enemy) and then uses a simple switch to pick between the three actions based on that state.

The merit of this idea to me is that it enforces the trichotomy on the level where the distinction between the behaviours actually manifests, which is literally just how the enemy chooses which of its behaviours it engages in. We already have the Bomb whose `engaged` behaviour is the same as its `aware` behaviour, and we're just now discussing a mimic-like enemy in the art-dev channel who could very well have the same `DoNothing` behaviour for both the `idle` and `aware` states.